producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,317 @@
+		return
+	}
+	select {
+	case w.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (w *consumerWriterImpl) Init() {
+	w.closeWG.Add(2)
+	go w.readAcksForever()
+	go w.resetConnectionForever()
+}
+
+func (w *consumerWriterImpl) readAcksForever() {
+	defer w.closeWG.Done()
+
+	var acks msgpb.Ack
+	for !w.isClosed() {
+		w.decodeLock.Lock()
+		err := w.encdec.Decode(&acks)
+		w.decodeLock.Unlock()
+		if err != nil {
+			w.m.decodeError.Inc(1)
+			w.signalReset()
+			// Adding some delay here to avoid this being retried in a tight loop
+			// when underlying connection is misbehaving.
+			time.Sleep(w.opts.AckErrorRetryDelay())
+			continue
+		}
+		for _, m := range acks.Metadata {
+			if err := w.router.Ack(metadataFromProto(m)); err != nil {
+				w.m.ackError.Inc(1)
+				w.logger.Errorf("could not ack metadata, %v", err)
+			}
+		}
+		// NB(cw) The proto needs to be cleaned up because the gogo protobuf
+		// unmarshalling will append to the underlying slice.
+		acks.Metadata = acks.Metadata[:0]
+	}
+}
+
+func (w *consumerWriterImpl) resetConnectionForever() {
+	defer w.closeWG.Done()
+
+	for {
+		select {
+		case <-w.resetCh:
+			w.resetWithConnectFn(w.connectWithRetry)
+		case <-w.doneCh:
+			return
+		}
+	}
+}
+
+func (w *consumerWriterImpl) resetWithConnectFn(fn connectFn) error {
+	w.m.resetConn.Inc(1)
+	conn, err := fn(w.addr)
+	if err != nil {
+		w.m.resetError.Inc(1)
+		w.logger.Errorf("could not connect to %s, %v", w.addr, err)
+		return err
+	}
+	w.reset(conn)
+	return nil
+}
+
+func (w *consumerWriterImpl) connectWithRetry(addr string) (net.Conn, error) {
+	continueFn := func(int) bool {
+		return !w.isClosed()
+	}
+	var (
+		conn net.Conn
+		err  error
+	)
+	fn := func() error {
+		conn, err = w.connectFn(addr)
+		return err
+	}
+	if attemptErr := w.retrier.AttemptWhile(
+		continueFn,
+		fn,
+	); attemptErr != nil {
+		return nil, fmt.Errorf("failed to connect to address %s, %v", addr, attemptErr)
+	}
+	return conn, nil
+}
+
+func (w *consumerWriterImpl) connectOnce(addr string) (net.Conn, error) {
+	conn, err := net.DialTimeout("tcp", addr, w.opts.DialTimeout())
+	if err != nil {
+		w.m.connectError.Inc(1)
+		return nil, err
+	}
+	tcpConn := conn.(*net.TCPConn)
+	if err = tcpConn.SetKeepAlive(true); err != nil {
+		w.m.setKeepAliveError.Inc(1)
+	}
+	return tcpConn, nil
+}
+
+func (w *consumerWriterImpl) reset(conn net.Conn) {
+	w.Lock()
+	defer w.Unlock()
+	// Close wakes up any blocking encode/decode function on the encodeDecoder.
+	// We are doing this mostly for the decode function.
+	// After which we could acquire the encode lock and decode lock.
+	w.encdec.Close()
+
+	w.decodeLock.Lock()
+	defer w.decodeLock.Unlock()
+	w.encodeLock.Lock()
+	defer w.encodeLock.Unlock()
+
+	// NB(cw): Must set the conn with both encode lock and decode lock
+	// to avoid data race.
+	w.encdec.Reset(conn)
+	w.lastResetNanos = w.nowFn().UnixNano()
+	w.cleanUpResetChannel()
+}
+
+func (w *consumerWriterImpl) cleanUpResetChannel() {
+	for {
+		select {
```

Comment:
If there's temporarily no signal on the reset channel, this function immediately returns, is that ok?

Reply:
Yeah that is fine, this function is just making sure we clean up the reset channel